gpytorch/models/pyro/pyro_gp.py

```diff
@@ -0,0 +1,114 @@
+    See `the Pyro examples <examples/09_Pyro_Integration/index.html>`_ for detailed examples.
+
+    Args:
+        :attr:`variational_strategy` (:obj:`~gpytorch.variational.VariationalStrategy`):
+            The variational strategy that defines the variational distribution and
+            the marginalization strategy.
+        :attr:`likelihood` (:obj:`~gpytorch.likelihoods.Likelihood`):
+            The likelihood for the model
+        :attr:`num_data` (int):
+            The total number of training data points (necessary for SGD)
+        :attr:`name_prefix` (str, optional):
+            A prefix to put in front of pyro sample/plate sites
+        :attr:`beta` (float - default 1.):
+            A multiplicative factor for the KL divergence term.
+            Setting it to 1 (default) recovers true variational inference
+            (as derived in `Scalable Variational Gaussian Process Classification`_).
+            Setting it to anything less than 1 reduces the regularization effect of the model
+            (similarly to what was proposed in `the beta-VAE paper`_).
+
+    Example:
+        >>> class MyVariationalGP(gpytorch.models.PyroGP):
+        >>>     # implementation
+        >>>
+        >>> # variational_strategy = ...
+        >>> likelihood = gpytorch.likelihoods.GaussianLikelihood()
+        >>> model = MyVariationalGP(variational_strategy, likelihood, train_y.size())
+        >>>
+        >>> optimizer = pyro.optim.Adam({"lr": 0.01})
+        >>> elbo = pyro.infer.Trace_ELBO(num_particles=64, vectorize_particles=True)
+        >>> svi = pyro.infer.SVI(model.model, model.guide, optimizer, elbo)
+        >>>
+        >>> # Optimize variational parameters
+        >>> for _ in range(n_iter):
+        >>>    loss = svi.step(train_x, train_y)
+
+    .. _Scalable Variational Gaussian Process Classification:
+        http://proceedings.mlr.press/v38/hensman15.pdf
+    .. _the beta-VAE paper:
+        https://openreview.net/pdf?id=Sy2fzU9gl
+    """
+    def __init__(self, variational_strategy, likelihood, num_data, name_prefix="", beta=1.0):
+        super().__init__(variational_strategy)
+        self.name_prefix = name_prefix
+        self.likelihood = likelihood
+        self.num_data = num_data
+        self.beta = beta
+
+    def guide(self, input, target, *args, **kwargs):
+        """
+        Gude function for Pyro inference.
+        Includes the guide for the GP's likelihood function as well.
+
+        Args:
+            :attr:`input` (`torch.Tensor`):
+                :math:`\mathbf X` The input values values
+            :attr:`target` (`torch.Tensor`):
+                :math:`\mathbf y` The target values
+            :attr:`*args`, :attr:`**kwargs`:
+                Additional arguments passed to the likelihood's `forward` function.
+        """
+        # Hack for getting correct sampling shape
+        pyro.sample("__throwaway__", pyro.distributions.Normal(0, 1)).shape
```

Comment:
what is this hack for?

Reply:
We're computing the log-likelihood term with MC-integratoin, which requires drawing samples from q(f). However, we don't want to draw samples from q(f) using `pyro.sample` since this would add E(q(f)) as a term in the loss function. Sampling from the throwaway distribution then allows us to get the sampling shape we would've gotten if we had called `pyro.sample` on q(f), and then we can draw non-pyro samples from q(f) by calling `function_dist(sample_shape)`.

Reply:
use pyro.sample but enclose it in pyro.block instead? won't the throwaway log prob leak into your loss?

Reply:
> won't the throwaway log prob leak into your loss?

It's a sample from a N(0, 1) distribution, so there's no parameters and it's pretty cheap to compute. There's also a `__throwaway__` sample in the guide so the KL component should exactly equal zero.

> use pyro.sample but enclose it in pyro.block instead

I tried `pyro.block` before, but unfortunately it doesn't respect the existing plate structure (i.e. we only end up with a single function_sample as a result).

Reply:
This is now fixed with the new interface (which is pretty finalized).